crates/e2e-test-utils/src/setup_builder.rs

```diff
@@ -0,0 +1,222 @@
+//! Builder for configuring and creating test node setups.
+//!
+//! This module provides a flexible builder API for setting up test nodes with custom
+//! configurations through closures that modify `NodeConfig` and `TreeConfig`.
+
+use crate::{node::NodeTestContext, wallet::Wallet, NodeBuilderHelper, NodeHelperType, TmpDB};
+use reth_chainspec::EthChainSpec;
+use reth_engine_local::LocalPayloadAttributesBuilder;
+use reth_node_builder::{
+    EngineNodeLauncher, NodeBuilder, NodeConfig, NodeHandle, NodeTypes, NodeTypesWithDBAdapter,
+    PayloadAttributesBuilder, PayloadTypes,
+};
+use reth_node_core::args::{DiscoveryArgs, NetworkArgs, RpcServerArgs};
+use reth_provider::providers::BlockchainProvider;
+use reth_rpc_server_types::RpcModuleSelection;
+use reth_tasks::TaskManager;
+use std::sync::Arc;
+use tracing::{span, Level};
+
+/// Type alias for tree config modifier closure
+type TreeConfigModifier =
+    Box<dyn Fn(reth_node_api::TreeConfig) -> reth_node_api::TreeConfig + Send + Sync>;
+
+/// Type alias for node config modifier closure
+type NodeConfigModifier<C> = Box<dyn Fn(NodeConfig<C>) -> NodeConfig<C> + Send + Sync>;
+
+/// Builder for configuring and creating test node setups.
+///
+/// This builder allows customizing test node configurations through closures that
+/// modify `NodeConfig` and `TreeConfig`. It avoids code duplication by centralizing
+/// the node creation logic.
+pub struct E2ETestSetupBuilder<N, F>
+where
+    N: NodeBuilderHelper,
+    F: Fn(u64) -> <<N as NodeTypes>::Payload as PayloadTypes>::PayloadBuilderAttributes
+        + Send
+        + Sync
+        + Copy
+        + 'static,
+    LocalPayloadAttributesBuilder<N::ChainSpec>:
+        PayloadAttributesBuilder<<N::Payload as PayloadTypes>::PayloadAttributes>,
+{
+    num_nodes: usize,
+    chain_spec: Arc<N::ChainSpec>,
+    is_dev: bool,
+    tree_config: reth_node_api::TreeConfig,
+    attributes_generator: F,
+    connect_nodes: bool,
+    tree_config_modifier: Option<TreeConfigModifier>,
+    node_config_modifier: Option<NodeConfigModifier<N::ChainSpec>>,
```

Comment:
I think this is a good step in the right direction, if needed we can iterate on this a bit more